File: framework/languages/dotnet/e2e/PubSub/Client/Scenarios/PsB2PublisherRestartScenario.cs
// Verifies PS-B2 Publisher Restart behavior.
using System.Diagnostics;
using PubSub.Client.Support;
using PubSub.Shared;
using Systems.Zlink.Stream.Connector.Contracts;
using Zlink.HttpClient;

namespace PubSub.Client.Scenarios;

// PS-B2: verifies that subscribers receive new events after the publisher process restarts.
internal static class PsB2PublisherRestartScenario
{
    public static async Task<Process> RunAsync(
        ZLinkHttpClient publisher,
        IReadOnlyList<ZLinkHttpClient> subscribers,
        ServerProcessLauncher processes)
    {
        var runId = Guid.NewGuid().ToString("N");

        // Establish a baseline delivery before stopping the publisher process.
        await publisher.Post("/publish/event")
            .Query("topic", PubSubNames.MainTopic)
            .Query("runId", runId)
            .Query("sequence", "1")
            .Query("value", "before-publisher-restart")
            .AsyncRaw();
        await WaitForSubscribersAsync(subscribers, new EvidenceWaitReq(
            ["event|", $"run={runId}", $"topic={PubSubNames.MainTopic}", "seq=1"],
            []));
        // Classic fanout registers no location store (config-3): a normal
        // replacement reaches terminal Drained, and the subscribers observe the
        // publisher going away through their own socket disconnect evidence
        // rather than a store row removal.
        var drain = (await publisher.Post("/admin/drain")
            .Timeout(TimeSpan.FromSeconds(35))
            .Async<DrainResultRes>()).Body;
        ZlinkStreamAssert.Ensure(
            drain.Result == nameof(Zlink.Framework.Contracts.Configuration.ZLinkFrameworkTerminationOutcome.Stopped),
            $"PS-B2 expected terminal Stopped, got {drain.Result}:{drain.Reason}.");

        await publisher.Post("/shutdown").AsyncRaw();
        await StateObservation.WaitUntilAsync(
            async () =>
            {
                try
                {
                    return (await publisher.Get("/health").AsyncRaw()).Status != 200;
                }
                catch (Exception ex) when ((ex is HttpRequestException || ex.InnerException is HttpRequestException))
                {
                    return true;
                }
            },
            "PS-B2 expected publisher to stop before restart.");
        // A publish while the process is down should fail at the HTTP boundary.
        await ZlinkStreamAssert.ExpectFailureAsync(async cancellationToken =>
            _ = await publisher.Post("/publish/event")
                .Query("topic", PubSubNames.MainTopic)
                .Query("runId", runId)
                .Query("sequence", "2")
                .Query("value", "during-publisher-down")
                .AsyncRaw(cancellationToken));

        // Restart the same publisher role and wait for the health endpoint before measuring recovery.
        var restartedPublisher = processes.StartPublisher();
        await StateObservation.WaitUntilAsync(
            async () =>
            {
                try
                {
                    return (await publisher.Get("/health").AsyncRaw()).Status == 200;
                }
                catch (Exception ex) when ((ex is HttpRequestException || ex.InnerException is HttpRequestException))
                {
                    return false;
                }
            },
            "PS-B2 expected restarted publisher to become healthy.");
        // A publisher health response does not mean that every existing fanout
        // subscription has completed its new transport handshake. Classic
        // fanout has no public connection-status surface, so use the existing
        // bounded delivery barrier to wait for each subscriber's first valid
        // post-restart record before publishing the measured marker.
        await Task.WhenAll(subscribers.Select((subscriber, index) =>
            FanoutReadiness.WaitUntilReceivingAsync(
                publisher,
                subscriber,
                $"PS-B2 subscriber {index + 1} after publisher restart")));

        // The subscribers keep their subscriptions; this second marker is the
        // scenario-visible recovery evidence after all subscriptions are ready.
        var readyRun = $"ready-{Guid.NewGuid():N}";
        await publisher.Post("/publish/event")
            .Query("topic", PubSubNames.MainTopic)
            .Query("runId", readyRun)
            .Query("sequence", "0")
            .Query("value", "publisher-restarted")
            .AsyncRaw();
        await Task.WhenAll(subscribers.Select(subscriber =>
            SubscriberObservation.WaitForEventAsync(subscriber, readyRun, 0)));

        // Every subscriber received the readiness probe before this measurement.
        await publisher.Post("/publish/event")
            .Query("topic", PubSubNames.MainTopic)
            .Query("runId", runId)
            .Query("sequence", "3")
            .Query("value", "after-publisher-restart")
            .AsyncRaw();

        // Recovery is proven by this exact first post-readiness event at every subscriber.
        await WaitForSubscribersAsync(subscribers, new EvidenceWaitReq(
            [],
            [])
        {
            ContainsAllLineGroups =
            [
                ["event|", $"run={runId}", $"topic={PubSubNames.MainTopic}", "seq=3|"]
            ]
        });
        Console.WriteLine("scenario PS-B2 passed");
        return restartedPublisher;
    }

    private static async Task WaitForSubscribersAsync(
        IReadOnlyList<ZLinkHttpClient> subscribers,
        EvidenceWaitReq request)
    {
        var waits = subscribers
            .Select(subscriber => subscriber.Post("/evidence/wait")
                .Body(request)
                // The server may use the full evidence wait window. Keep the HTTP
                // deadline slightly longer so a valid terminal response is observable.
                .Timeout(TimeSpan.FromMilliseconds(request.TimeoutMilliseconds + 2000))
                .Async<string[]>()
                .AsTask())
            .ToArray();
        await Task.WhenAll(waits);
    }
}

File: framework/languages/node/e2e/PubSub/Client/Scenarios/ps-b2-publisher-restart-scenario.ts
// PS-B2: Publisher 재시작 뒤 기존 subscriber가 새 event를 받는다 시나리오를 검증한다.
import { randomUUID } from 'node:crypto';
import { PubSubNames } from '../../Shared/messages';
import { getStatus, postJsonWithin, postStatus } from '../../../http-client';
import { ensure, eventually, isConnectionFailure } from '../Support/scenario-assert';
import {
  waitForEvent
} from '../Support/subscriber-observation';
import { ServerProcessLauncher, type DynamicProcess } from '../Support/server-process-launcher';
import { publishEvent, publishUntilAllDelivered } from './ps-a1-fanout-basic-delivery-scenario';

export async function runPsB2(
  publisher: string,
  subscribers: readonly string[],
  processes: ServerProcessLauncher
): Promise<DynamicProcess> {
  const runId = randomUUID().replaceAll('-', '');
  await publishEvent(publisher, PubSubNames.mainTopic, runId, 1, 'before-publisher-restart');
  await Promise.all(subscribers.map((url) => waitForEvent(url, runId, 1, 'before-publisher-restart')));

  const drain = await postJsonWithin<{ readonly kind: string; readonly reason?: string }>(
    publisher, '/admin/drain', {}, 35_000
  );
  ensure(
    drain.kind === 'drained' && drain.reason === undefined,
    `PS-B2 expected terminal Drained, got ${drain.kind}/${drain.reason ?? ''}.`
  );
  await postStatus(`${publisher}/shutdown`);
  await eventually(async () => {
    try {
      return await getStatus(`${publisher}/health`) !== 200;
    } catch (error) {
      return isConnectionFailure(error);
    }
  }, 'PS-B2 expected publisher process to stop.');

  try {
    await publishEvent(publisher, PubSubNames.mainTopic, runId, 2, 'during-publisher-down');
    throw new Error('PS-B2 expected publish attempt to fail while publisher is down.');
  } catch (error) {
    ensure(isConnectionFailure(error), 'PS-B2 expected HTTP connection failure while publisher is down.');
  }

  const restarted = processes.startPublisher();
  await restarted.waitReady();
  await publishUntilAllDelivered(
    publisher, subscribers, PubSubNames.mainTopic, 3, 'after-publisher-restart'
  );
  console.log('scenario PS-B2 passed');
  return restarted;
}
